### ISSUE TITLE
How far back to do we go in supporting old IE by fixing its bugs?

### ISSUE THREAD

**Issue by jensimmons** [mozdevs/cssremedy#3]
Many of projects similar to this have been removing their code that fixes bugs in really old versions of Internet Explorer. 

This code for instance:
```
article, aside, figcaption, figure, footer, header, hgroup, main, nav, section {
  display: block;
}
```
Do we bother? 

How far back to we want to "support"?

**Comment by scottkellum** [comment]
IE11 and no further back would be what I would support for mass market projects. People are still stuck on it at work, but most of them have been upgraded to the latest version at this point.

**Comment by callahad** [comment]
IE11 only.

Microsoft is killing IE10 [next January](https://www.computerworld.com/article/3337528/web-browsers/microsoft-to-end-ie10-support-years-early-tells-enterprises-to-upgrade-to-ie11.html), and it would be irresponsible for us to support a browser that's insecure and unsupported by its own vendor.

**Comment by jensimmons** [comment]
I find it hard to ignore IE10 and before. 

I just added: 
```
img {
  border-style: none; /* Remove the border on images inside links in IE 10 and earlier. */
}
```

It's not that I think we should recommend people use IE10 (and earlier). We should not. It's that people _are_ using it, and well, maybe websites should do a few easy things to make it less of a horrible experience. Like removing that ugly blue border from images that have links on them.

Conversely, maybe we _should_ encourage sites to look ugly and awful in old IE. That way people using IE6-10 will realize, "OMG, this browser is old and crummy — look how awful everything is."

>it would be irresponsible for us to support a browser that's insecure and unsupported

I agree with this. And yet, I'm not sure I see adding code like what above as "supporting an insecure browser" as much as removing hassles for developers by fixing little things that _will_ come up for the users that _are_ using IE10-. I'm torn.

On the web, you don't get to choose what hardware or software people use to get to your website.

**Comment by jensimmons** [comment]
Here's another example — from Normalize — of the kind of thing we would include, or not include:
```

 * 1. Add the correct box sizing in IE 10.
 * 2. Remove the padding in IE 10.
 */

[type="checkbox"],
[type="radio"] {
  box-sizing: border-box; /* 1 */
  padding: 0; /* 2 */
}
```

**Comment by jensimmons** [comment]
Using CanIUse to get at the global browser stats coming from StatsCounter, the current usage is:
IE11        2.37%
IE10       0.07%
IE9         0.18%
IE8         0.16%
IE7         0.01%
IE6         0.01%

Interesting that IE 8 & 9 have far more usage than IE10. Which means "supporting" 
- IE11-only makes sense, 
- IE8-11 makes sense, 
- but IE 10-11 does not make any sense. 

From StatsCounter:
Blackberry    0.02%
IE6-10           0.43%
Android         1.15%     (Android 2-4.4  0.81%)
Edge              2.13%
IE11                2.37%
Opera            2.91%    (Opera Mini 1.81%)
Samsung       3.64%
UC                 3.83%

As always — it's funny how developers obsess, still, over old IE, while completely disregarding a pile of mobile-OS-based browsers that have far more market share.

**Comment by jensimmons** [comment]
Probably the biggest decision is whether or not to include:
```
article, aside, figcaption, figure, footer, header, hgroup, main, nav, section {
  display: block;
}
```
to get the HTML5 structural elements to display correctly in IE 8 – 10.

**Comment by frivoal** [comment]
> Probably the biggest decision

I think I would, as if you don't the site goes a bit beyond looking ugly, and starts going into the impossible to understand territory.

**Comment by sandren** [comment]
I'm kind of torn.

On the one hand, if the goal of this project is to set properties and values as if "the CSSWG were creating the CSS today", then we know for certain that this would be their intention:

```
article, aside, figcaption, figure, footer, header, hgroup, main, nav, section {
  display: block;
}
```

But on the other hand, people that are using browsers that are over half a decade old must be so used to seeing broken websites, that at this point they likely cannot even tell the difference between what is broken and what is not anymore.

I am a huge advocate for web accessibility, but I don't think dropping support for IE necessarily violates that endeavor. After all, we don't provide versions of legal disclaimers in Old English. 😅

**Comment by eallenOP** [comment]
You absolutely have to draw the line somewhere, based on some criteria. There are even older browsers that are not being considered here (and possibly some that are even the same age as earlier IE) but whether the (arbitrary) line is drawn at a certain age, or a certain percentage use, or some other criteria, it needs to be applied consistently and decisively. It would be a bit odd to make that line at "we support old stuff that isn't in use much but only Microsoft".

Some options:
- don't support browsers officially dropped before `date`
- don't support browsers with less than `percentage` market share
- a combination of the above two
- support old/unusual browsers but only if it takes less than `number` lines of code (or move it to another file?)

I totally agree with
>As always — it's funny how developers obsess, still, over old IE, while completely disregarding a pile of mobile-OS-based browsers that have far more market share.

**Comment by Sora2455** [comment]
My two cents: we shouldn't be babying the users of outdated browsers, we should be telling them to get off the internet:
```
body:after {
    content: "Your browser is outdated, and will suffer a degraded experience on this website. Please upgrade to a newer browser such as Firefox.";
    position: fixed;
    left: 0;
    bottom: 0;
    width: 100%;
    background-color: red;
    color: white;
    text-align: center;
    font-size: 20px;
    font-size: 1.25rem;
    line-height: 1.5;
}
@supports(display: grid){
    body:after {
        display: none;
    }
}
@media only screen and (min-width: 768px) {
    body:after {
        font-size: 32px;
        font-size: 2rem;
    }
}
```
(I picked Firefox because the ESR still supports XP, so users have no excuse.)

**Comment by jensimmons** [comment]
Looking up numbers on Stats Counter in Sept 2019.....

IE, globally, every version: 1.97%
IE11, globally: 1.68%
Which puts IE6-10 at 0.29%.
3.2 billion people online.

3.2 billion * 0.29% is 9,280,000 people.

**Comment by scottkellum** [comment]
Maybe support doesn’t have to be a binary? Some of the low hanging fruit you bring up makes a lot of sense to include. At the same time, saying a browser is supported is scary because that means some degree of quality assurance on those browsers.

**Comment by mirisuzanne** [comment]
Yeah, I think we've landed on a rough policy for now:

- We will add proven fixes for pre-IE11 browsers when they are light-weight and non-invasive. 
- We won't "support" pre-IE11 browsers with any sense of urgency or priority.
- We're not trying to polyfill new CSS features like grid, or the full html5 "shiv"

**Comment by jensimmons** [comment]
I agree with Scott's comment. We don't have to handle this as a binary. Miriam & I just had a nice long conversation about it today, coming to consensus on an approach. 

The goal is to give Authors a helpful shortcut to fewer catastrophic bugs in IE6-10. The goal is not to get all the old browsers to work identically to modern browsers (an impossible task). 

Here's an example of what success would look like:
- A team decides "not to support" IE. 
- They use CSS Remedy, without thinking about the IE bits. And launch.
- Users who use IE go to the website, despite the team's lack of IE "support". 
- The website loads. Because it's the web. This isn't Android/iOS/macOS/Windows. 
- There are fewer problems with the site, because Remedy is helping out.

Now, the goal is not to fix all the bugs. Nope. Can't. Think about the difference between these two:

1. We prevent the newer HTML5 elements from being `display: inline`, by including `article, etc {display:block}`. Which prevents total layout chaos, and gets everything into a basic flow, as expected by all parties. 
2. We get deep into tracking down newly-discovered lack-of-interop bugs with IE, test them, and implement fixes to get padding, margin, etc to try and act in an identical fashion to a modern browser. Which results in a bit of polish in the small details of the layout. 

I want us to do the first. I don't want us to do the second. 

I do think a lot of work has already been done (in other reset projects, like Normalize) — and it's easy enough for us to copy that work into this project, without spending a lot of time testing or researching. 

Some may argue that we shouldn't add the extra kilobytes, but I disagree. It won't be that much code. There are many other places a project should trim. This will add up to far less than one tiny image. 

Some have argued (see above) that including "support" for IE6-10 will act as an endorsement of those browsers, and encourage users to use them. I disagree. First, the users of the websites who use Remedy will not be reading our code — so that's not going to influence them. And development teams will make their own decision based on their preferences & their Q/A testing resources, not based on Remedy. Second, users of IE are already having a terrible experience, making it slightly less terrible on some sites is not going to be motivation enough for them to keep using it when they otherwise would stop. Usually those users have no reasonable choice and they have to use IE for some reason. Why would we want to prevent them from doing what they need to do? That seems wrong to me. Especially if they are trying to do something important like applying for a job, or submitting healthcare paperwork. We should support users over theoretical purity. 

So I say, let's carry over the IE6-11 bug fixes from Normalize, etc. Let's put a bit of time into being thoughtful about it... merging together the best of those fixes. Let's not put a disproportionate amount of time researching or testing these fixes. Or debating whether or not it's the right call. And if discover other / new additions to the IE fixes, let's prioritize the ones that have a big impact, and not worry so much about the ones that polish up tiny details. 

Good enough is good enough. Let's put our time & attention on the parts for the more modern browsers.

**Comment by thierryk** [comment]
> We prevent the newer HTML5 elements from being display: inline, by including article, etc {display:block}. Which prevents total layout chaos, and gets everything into a basic flow, as expected by all parties.

No, you don't!
Because the issue is **not limited to `display` alone**.
In oldIE, one needs a bit of JS on top of that rule to make things work. Something like this for example (for `main`):

```
<!--[if IE]><script>document.createElement("main");</script><![endif]-->
```

In other words, `article, etc {display:block}` does nothing unless the author of the document is using `createElement()` and if they do, then I'm sure they know about specifying `display:block` for the HTML5 elements they are using.  

I see 3 options:

   * Adding `details` (which is missing) to that rule **and adding a comment mentioning the need for `createElement()`**
   * Replacing that rule with `header, details {display:block}` to cater for Opera Mini
   * Removing that rule entirely

I doubt that ppl using `createElement()` rely on *CSSRemedy* for the added styling they need, but I could be wrong.

**Comment by mirisuzanne** [comment]
@thierryk You missed the point of Jen's comment. This is a discussion of our _planned approach_ to legacy browsers, with hypothetical examples, not a claim about what's already done & finalized. The point is that we _may provide_ some CSS basics that help legacy browsers, but our goal is _not_ full testing or "support" of those browsers. There is no reason to challenge or contradict any of that based on what we have so far. We're still very much in the early stages, and state very clearly in our README that this is not a "stable" project. 

If you want to open an issue for improving on our html5-tag remedy, you are welcome to. There are many open discussions about the best approach to specific features. 

But this is not the proper thread, or a proper way to frame it.

**Comment by thierryk** [comment]
@mirisuzanne before opening an issue I wanted to know what the intentions of this project are exactly, hence why I'm commenting on an existing issue labelled "philosophical", and why I mentioned 3 options:
 
   * this project wants to "fix" old IE (a comment would be added to help users make sense of the limitation)
   * this project wants to fix *recent* browsers that don't support  HTML5 elements (ie Opera Mini)
   * this project does not offer support for new HTML5 elements (supporting oldIE would require more than that rule, we don't support Opera Mini as its global usage is only 1.16%, etc.)
    
I believe most people would not know about HTML5 elements and oldIE so I think it is important that comments, or lack of comments, in this project do not lead people think that `display:block` is all they need to support oldIE.

In other words, I am not here to challenge or contradict anybody. And my comments are not about what this project should do or be either. I'm just sharing with you how I think this project should convey information to its users. Because to me, the styles sheet and this thread suggest that `display:block` will prevent breakage in old browsers (which is not the case).

**Comment by mirisuzanne** [comment]
The philosophical resolution of this thread is:
 
1. Modern browsers are the top priority
2. Remedies for legacy-browsers can be considered **case-by-case** depending on their complexity & impact, but are not the top priority (open an issue)
3. We don't "support" a specific set of browsers with a specific set of features

That doesn't cleanly map to your three categories, but it does leave individual cases open to discussion. You're welcome to open an issue.

**Comment by thierryk** [comment]
[Done!](https://github.com/jensimmons/cssremedy/issues/79)
